### PR TITLE
Inherited test methods don't support their annotations (fixes #1032)

### LINKED ISSUE #1032: 🐞: Inherited test methods don't support their annotations.
### What happened?

The `getTestMethod` method in the `io.qameta.allure.junitplatform.AllureJunitPlatformUtils` class doesn't function as expected for tests that are inherited. This is because it uses `aClass.getDeclaredMethods()`, which only operates on the current instance and does not include methods that are inherited. This can cause issues with annotations related to methods, such as descriptions and others.

### What Allure Integration are you using?

allure-junit-platform

### What version of Allure Integration you are using?

2.27.0

### What version of Allure Report you are using?

2.27.0

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ingvard**
fixes #1032 

### Context
The `getTestMethod` method in the `io.qameta.allure.junitplatform.AllureJunitPlatformUtils` class doesn't function as expected for tests that are inherited. This is because it uses `aClass.getDeclaredMethods()`, which only operates on the current instance and does not include methods that are inherited. This can cause issues with annotations related to methods, such as descriptions and others.

### Revision 1
In this revision, I used the `aClass.getMethods()` that returns all public methods on a checking class.

### Revision 2
However, revision 1 covers one of the test cases when all test methods are public. It doesn't work when:
1. Methods have package-based visibility that is usually for unit tests.
```
   public abstract static class ParentTest implements GrandparentTest {
        @Test
        @Description(TEST_DESCRIPTION)
        @Story(INHERITED_TEST_PARENT_STORY)
        @Link(TEST_LINK)
        void parentTest() {}
    }
 ```
2. Methods are members of the interface (as default methods).
```
 public interface GrandparentTest {
        @Test
        @Description(TEST_DESCRIPTION)
        @Story(INHERITED_TEST_GRANDPARENT_STORY)
        @Link(TEST_LINK)
        default void grandparentTest() {}
    }
```
3. Parent class has an upper-class annotation that must be applied to all inheritors.
```
 @Epic(INHERITED_TEST_EPIC)
    @Feature(INHERITED_TEST_FUTURE)
    public interface GrandparentTest { }
```

In short, I have to implement two methods placed into `ReflectionUtils` to recursively collect method and annotation information. These methods are designed as simplified logic of `org.springframework.data.util.ReflectionUtils` specifically for our needs.

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

**Review by baev — APPROVED** [review]

File: allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+public final class ReflectionUtils {
```
Comment (baev) [inline allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java:28]: add javadoc

File: allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    private ReflectionUtils() {
```
Comment (baev) [inline allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java:29]: add `throw new IllegalStateException("Do not instance");`

File: allure-java-commons/src/main/java/io/qameta/allure/util/AnnotationUtils.java [hunk 3]
```diff
@@ -116,7 +117,7 @@ public static Set<Link> getLinks(final Collection<Annotation> annotations) {
      * @return discovered labels.
      */
     public static Set<Label> getLabels(final AnnotatedElement annotatedElement) {
-        return getLabels(annotatedElement.getAnnotations());
+        return getLabels(getAllAnnotations(annotatedElement));
```
Comment (baev) [inline allure-java-commons/src/main/java/io/qameta/allure/util/AnnotationUtils.java:120]: should be added to getLinks method as well

Comment (baev) [reply allure-java-commons/src/main/java/io/qameta/allure/util/AnnotationUtils.java:120]: add test

File: allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java (hunk 1 above)
Comment (baev) [reply allure-java-commons/src/main/java/io/qameta/allure/util/ReflectionUtils.java:28]: add tests

**Review by baev — CHANGES_REQUESTED** [review]

**Review by baev — CHANGES_REQUESTED** [review]
Found a bug:

```java
public interface First {

    default void a() {
    }
}
public interface Second extends First {

    default void b() {
    }
}
```

then `ReflectionUtils.getAllDeclaredMethods(Second.class)` will produce duplicate results:

```
public default void Second.b(),
public default void First.a(),
public default void First.a()
```

**Comment by baev** [comment]
fixed by #1317
